### PR TITLE
Feat/tunnelv5

### PR DESCRIPTION
**Pull request by wussler**

### CODE REVIEW

File: lib/key/utils.js [hunk 1]
```diff
@@ -2,11 +2,14 @@ import { openpgp } from '../openpgp';
...
 // returns promise for generated RSA public and encrypted private keys
-export const generateKey = (options) => {
+export async function generateKey(options) {
     options.date = typeof options.date === 'undefined' ? serverTime() : options.date;
     return openpgp.generateKey(options);
-};
-export const generateSessionKey = (algorithm) => openpgp.crypto.generateSessionKey(algorithm);
+}
+
+export async function generateSessionKey(algorithm) {
```
Comment (dhoko) [inline lib/key/utils.js]:
```suggestion
export const generateSessionKey = openpgp.crypto.generateSessionKey;
```

File: lib/key/utils.js [hunk 2]
```diff
@@ -101,3 +104,67 @@ export async function cloneKey(inputKey) {
...
+ *
+ * @param {Array<Options>}         Private key d, public part of ECDHE V, fingerprint and curve (name or OID)
+ * @returns {Promise<Uint8Array>}  Generated ephemeral secret
+ * @async
+ */
+export async function genPrivateEphemeralKey(options) {
+    const curveObj = new openpgp.crypto.publicKey.elliptic.Curve(options.Curve);
+    const oid = new openpgp.OID(curveObj.oid);
+    const S = await openpgp.crypto.publicKey.elliptic.ecdh.genPrivateEphemeralKey(curveObj, options.V, options.d);
+    const cipher_algo = openpgp.enums.read(openpgp.enums.symmetric, curveObj.cipher);
```
Comment (dhoko) [inline lib/key/utils.js]: Can you change the format of var to `cipherAlgo` ;)

File: lib/key/utils.js [hunk 3]
```diff
@@ -101,3 +104,67 @@ export async function cloneKey(inputKey) {
...
+        false
+    );
+
+    return {V, Z}
+}
+
+/**
+ * Generate ECDHE secret from private key and public part of ephemeral key
+ *
+ * @param {Array<Options>}         Private key d, public part of ECDHE V, fingerprint and curve (name or OID)
```
Comment (dhoko) [inline lib/key/utils.js]: Wrong type, it's an Object.
Can you use destructuring to improve the readability of options ? (_it's easier to doc as we can do per options + types + what we need without looking at the body_)

File: lib/key/utils.js [hunk 4]
```diff
@@ -101,3 +104,67 @@ export async function cloneKey(inputKey) {
...
+ * Generate ECDHE key and secret from public key
+ *
+ * @param {Array<Options>}                              Public key Q, fingerprint and curve (name or OID)
+ * @returns {Promise<{V: Uint8Array, Z: Uint8Array}>}   Returns public part of ephemeral key and generated ephemeral secret
+ * @async
+ */
+export async function genPublicEphemeralKey(options) {
+    const curveObj = new openpgp.crypto.publicKey.elliptic.Curve(options.Curve);
+    const oid = new openpgp.OID(curveObj.oid);
+    const { V, S } = await openpgp.crypto.publicKey.elliptic.ecdh.genPublicEphemeralKey(curveObj, options.Q);
```
Comment (dhoko) [inline lib/key/utils.js]: No need to wrap it with try/catch ? Is it expected ?

File: lib/key/utils.js [hunk 5]
```diff
@@ -101,3 +104,67 @@ export async function cloneKey(inputKey) {
...
 }
+
+/**
+ * Generate ECDHE key and secret from public key
+ *
+ * @param {Array<Options>}                              Public key Q, fingerprint and curve (name or OID)
+ * @returns {Promise<{V: Uint8Array, Z: Uint8Array}>}   Returns public part of ephemeral key and generated ephemeral secret
+ * @async
+ */
+export async function genPublicEphemeralKey(options) {
```
Comment (dhoko) [inline lib/key/utils.js]: idem about destructuring (cf next comment)

**Comment by dhoko** [comment]
Can you add tests too ?

File: lib/key/utils.js (hunk 4 above)
Comment (wussler) [reply lib/key/utils.js]: It is try/catched upstream in the server

**Comment by wussler** [comment]
> Can you add tests too ?

Tests are done in Openpgpjs and Slim-API, I will add them soon

**Pull request merged**